### DISCUSSION TITLE
Snapshot of Config/ Config Versions

### DISCUSSION THREAD

**Discussion by pratikmishra356** [juspay/superposition#71] [Ideas]
Introducing new feature - Config Versions. We will now support version history for configs which included contexts, overrides and default configs.

**Comment by pratikmishra356** [comment]
### **Database Design**

    CREATE TABLE public.config_versions (
    id bigint PRIMARY KEY,
    config json NOT NULL,  -- stores config
    config_hash TEXT NOT NULL,  -- stores hash of config 
    tags varchar(100)[] check (array_position(tags, null) is null),  - user can tag any particular changes in the config
    created_at timestamp without time zone DEFAULT CURRENT_TIMESTAMP NOT NULL
     );

### **Api changes**
Apis need to be added/updated : 

Endpoint: /config/versions         → add
Method: GET
Description: metadata for version history.

Endpoint: /config?version=””         → update
Method: GET
Description: will take version in query params, if version not provided, it will give the latest config

***1** - This will accept another field `tags` of type `Option<Vec<String>>` by which user can tag the particular changes
Endpoint: /context
Method: PUT
Description: *1

Endpoint: /context/move/{ctx_id}         → update
Method: PUT
Description: *1

Endpoint: /context/{ctx_id}         → update
Method: DELETE
Description:  *1

Endpoint: /priority/recompute         → update
Method: DELETE
Description:  *1

Endpoint: /default-config/{key}         → update
Method: PUT
Description:  *1

Endpoint: /default-config/{key}         → update
Method: DELETE
Description:  *1


### **Client Changes**

While initiating cac_client we would require two extra parameters - 
    top_n_config   - integer  // defaults to 5, max restricted to 50
    filter_tags : array of string // filtered based on atleast any one in the given list should be present in table tags column
client functions   **eval, get_resolved_config, get_default_config**  will accept one more optional parameter **version**, if version not provided , it will assume the latest config from the stored configs

**Reply by Datron** [reply]
> client functions eval, get_resolved_config, get_default_config will accept one more optional parameter version, if version not provided , it will assume the latest config from the stored configs

@pratikmishra356 if version provided to these eval, get_resolved_config, get_default_config isn't present in the client, will we go to server to fetch the data?

**Reply by pratikmishra356** [reply]
yes we should fetch from server

**Comment by pratikmishra356** [comment]
Approach for Adding x-config-version-id in response headers: 

Following Apis will have this header in response : 
**Context** - create/update/delete/move
**Default-Configs** - create/update/delete
**Config** - All get requests


Implementation Approach : 
1. Adding in middleware by extracting method and route ( not sure how this will work)
2. Overriding responder trait for Json

**Reply by Datron** [reply]
@pratikmishra356 you mentioned that you are adding functions to each handler. Can you elaborate on that approach?

**Reply by pratikmishra356** [reply]
@Datron 
Problem Statement is we need to add headers for some group of apis

Approach is , that we will define a struct **CustomResp** with responder trait implementation , 
This struct will accept body and headers , and it will create an http response with the given body and headers.
In handler we can construct this struct and return just like Json struct

**Reply by Datron** [reply]
Sure, sounds good.
